docs/architecture/adr-047-handling-evidence-from-light-client.md

```diff
@@ -19,161 +20,193 @@ the light client detector receives two conflicting headers
 ```
 
-When a full node receives a `ConflictingHeadersTrace`, it should
-a) validate it b) figure out if malicious behaviour is obvious (immediately
-slashable) or run the amnesia protocol.
-
-### Validating headers
-
-Check headers are valid (`ValidateBasic`), are in order of ascending height, and do not exceed the `MaxTraceSize`.
-
-### Finding Block Bifurcation
-
-The node pulls the block ID's for the respective heights of the trace headers from its own block store.
-
-First it checks to see that the first header hash matches its first `BlockID` else it can discard it. 
+This approach has the advantage of not requiring as much processing on the light 
+client side in the event that an attack happens. Although, this is not a significant 
+difference as the light client would in any case have to validate all the headers 
+from both witness and primary. Using traces would consume a large amount of bandwidth
+and adds a ddos vector to the full node.
 
-If the last header hash matches the nodes last `BlockID` then it can also discard it on the assumption that a fork can not remerge and hence this is just a trace of valid headers. 
 
-The node then continues to loop in descending order checking that the headers hash doesn't match it's own blockID for that height. Once it reaches the height that the block ID matches the hash it then sends the common header, the trusted header and the diverged header (common header is needed for lunatic evidence) to determine if the divergence is a real offense to the tendermint protocol or if it is just fabricated.
-
-### Figuring out if malicious behaviour
-
-The node first examines the case of a lunatic attack:
-
-* The validator set of the common header must have at least 1/3 validator power that signed in the divergedHeaders commit
-
-* One of the deterministically derived hashes (`ValidatorsHash`, `NextValidatorsHash`, `ConsensusHash`,
-`AppHash`, or `LastResultsHash`) of the header must not match:
-
-* We then take every validator that voted for the invalid header and was a validator in the common headers validator set and create `LunaticValidatorEvidence`
-
-If this fails then we examine the case of Equivocation (either duplicate vote or amnesia):
+## Decision
 
-*This only requires the trustedHeader and the divergedHeader*
+The light client will be divided into two components: a `Verifier` (either sequential or 
+bisection) and a `Detector` (see [Informal's Detector](https://github.com/informalsystems/tendermint-rs/blob/master/docs/spec/lightclient/detection/detection.md))
+. The detector will take the trace of headers from the primary and check it against all 
+witnesses. For a witness with a diverging header, the detector will first verify the header
+by bisecting through all the heights defined by the trace that the primary provided. If valid,
+the light client will trawl through both traces and find the point of bifurcation where it
+can proceed to extract any evidence (as is discussed in detail later). 
```

Comment:
what if point of bifurcation cannot be found?

Reply:
It has to be. If there are different headers at height 100 and the same headers at height 10 at one point during bisection they must diverge